### ISSUE TITLE
value  and its explicit tag verification

### ISSUE THREAD

**Issue by ilyam8** [adrienverge/yamllint#248]
For syntax error `yamllint` uses `yaml.parse`

https://github.com/adrienverge/yamllint/blob/542ae758f5de70a37c6c322c16b83b92fdc008bc/yamllint/linter.py#L177-L179

It doesnt catch errors like:

```yaml
key: !!bool value
```

I agree that it is not syntax error, but i think it would be better to have full check and instead of `yaml.parse` do `yaml.parse(buffer, Loader=yaml.FullLoader)`.

___

My point that current yamllint check pass doesnt guarantee that the yaml file is the correct yaml file and can be loaded w/o errors.

**Comment by adrienverge** [comment]
That's a good point.

Unfortunately `yaml.FullLoader` has other side-effects, and the goal of yamllint is to read the yaml *text*, not to interpret it.

Maybe we could create a rule `consistent-types` to detect inconsistencies like:

```yaml
a: !!bool text
b: !!int true
c: !!str 123
d: !!binary {k: v}
```
?

**Comment by ilyam8** [comment]
> and the goal of yamllint is to read the yaml text, not to interpret it.

I may be wrong, but i think for a user "green" yamllint check means it is safe to merge. In fact it is not. At least it was the case for me. So main goal is to ensure that everything is good.


Btw is this the only difference between `yaml.parse` and `yaml.load` with `yaml.FullLoader`? Do you know what other problem `yaml.load` is able to raise but `yaml.parse` cant?


> rule consistent-types to detect inconsistencies like:

i guess this one shouldnt be optional and always be on, i see no use case you want to have it disabled.

> has other side-effects

What? What is side-effect if we change syntax_error to do load with FullLoader?

**Comment by perlpunk** [comment]
Regarding the types: Note that YAML 1.2 has different resolving rules than YAML 1.1 (for example it removes a lot of the boolean values, and integers and floats are matched differently).
PyYAML might support YAML 1.2 at some point, so yamllint would need to be prepared for configuring this too.

Regarding the example `key: !!bool value`: I haven't seen YAML files in the real world where those standard tags are used explicitly. What's the use-case?

**Comment by ilyam8** [comment]
@perlpunk 

>  haven't seen YAML files in the real world

Same.

My main point is

> Current yamllint check pass doesnt guarantee that the yaml file is the correct yaml file and can be loaded w/o errors.

I assume `yaml.load(buffer, Loader=yaml.FullLoader)` guarantees it is valid - no error on unmarshal. `yaml.parse` does not.

I am not aware if tag/value verification is the only case. If it is probably it is acceptable to ignore it.

**Comment by adrienverge** [comment]
`yaml.safe_load()` (`.load()` shouldn't be used) returns a Python object representing data, without any information on tokens. With this Python object (for example `{foo: bar}`), most yamlllint rules wouldn't have necessary information to work.
`yaml.parse()` returns a series of YAML tokens to read and check, this is what yamllint needs to do its checks.

Regarding the example `key: !!bool value`, I agree it's rare, but it's in the YAML spec. For example you can explicitly prevent strings from being parse to dates, like in `not-date: !!str 2020-04-08`. Also some users could want to write base64-encoded data like `!!binary AcgggoBADs=` and make sure it's not treated as a regular string.

@ilyam8 I think a new rule like `consistent-types` would fit your need, no? This rule would report errors like `key: !!bool value`.

**Comment by ilyam8** [comment]
> yaml.safe_load() (.load() shouldn't be used) returns a Python object representing data, without any information on tokens. 

i meant to use it just to validate configuration and not to use return value.

> I think a new rule like consistent-types would fit your need, no?

Yes, it does.

**Comment by perlpunk** [comment]
Related, might be useful:
I created a repo with test data for different YAML Schemas: https://perlpunk.github.io/yaml-test-schema/
You can see all regexes, and test data.
The test data is available as a YAML file, but the format might change.
Note that PyYAML doesn't implement the 1.1 types correctly.
